apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/FlushableConsumerBase.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// Base class for the <see cref="IFlushableConsumer{T}"/> that exposes a <see cref="SemaphoreSlim"/> to
+    /// protect the buffer.
+    /// </summary>
+    internal abstract class FlushableConsumerBase<T> : IFlushableConsumer<T>
+    {
+        /// <summary>A semaphore to protect the buffer.</summary>
+        protected readonly SemaphoreSlim Semaphore = new SemaphoreSlim(1);
+
+        /// <summary>The consumer to flush to.</summary>
+        protected readonly IConsumer<T> Consumer;
```

Comment:
I wouldn't have either of these protected fields - I'd let each subclass handle its own storage. The base class can *just* handle the semaphore, basically. It means a tiny bit of duplication, but really not a lot.